demo/lightning/checkpoint/simulated/multiprocessing_train.py

```diff
@@ -15,24 +15,120 @@
  """
 import argparse
 import os
-import time
 import statistics
-from typing import Optional, Dict
+import time
+from typing import Dict, Optional, TextIO
+
 
 import torch
-import torch.nn as nn
 import torch.distributed as dist
-import torch.multiprocessing as mp
 import torch.distributed.checkpoint as dist_cp
+import torch.multiprocessing as mp
+import torch.nn as nn
+
 from lightning.pytorch.strategies import FSDPStrategy
-from dataflux_pytorch.lightning.gcs_filesystem import GCSDistributedWriter, GCSDistributedReader
 
+from dataflux_pytorch.lightning.gcs_filesystem import GCSDistributedReader, GCSDistributedWriter
+
+# Constants for distributed setup
 MASTER_ADDR = 'localhost'
 MASTER_PORT = '12355'
 
 
+def write_model_structure(file: TextIO, model: nn.Module, indent: str = '') -> None:
+    """
+    Write the full model structure and state to a file.
+
+    This function writes a detailed representation of the model, including its
+    structure, state dictionary, and any dummy tensors if present.
+
+    Args:
+        model (nn.Module): The PyTorch model to be written to file.
+        filename (str): The name of the file to write the model information to.
+    """
+    for name, module in model.named_children():
+        file.write(f"{indent}{name}:\n")
+        if list(module.children()):
+            write_model_structure(module, file, indent + '  ')
+        else:
+            file.write(f"{indent}  {module}\n")
+            for param_name, param in module.named_parameters():
+                file.write(f"{indent}    {param_name}: {param.shape}\n")
+                file.write(f"{indent}      Values: {param.data}\n")
+
+
+def write_full_model(model: nn.Module, filename: str) -> None:
+    with open(filename, 'w') as f:
+        f.write("Model Structure:\n")
+        write_model_structure(model, f)
+        f.write("\nModel State Dict:\n")
+        for key, value in model.state_dict().items():
+            f.write(f"{key}:\n")
+            f.write(f"  Shape: {value.shape}\n")
+            f.write(f"  Values: {value}\n")
+
+            if hasattr(model, 'dummy_tensors'):
+                f.write("\nDummy Tensors:\n")
+                for i, tensor in enumerate(model.dummy_tensors):
+                    f.write(f"dummy_tensor_{i}:\n")
+                    f.write(f"  Shape: {tensor.shape}\n")
+                    f.write(f"  Values: {tensor}\n")
+
+
+def write_state_dict(state_dict: Dict[str, torch.Tensor], filename: str) -> None:
+    with open(filename, 'w') as f:
+        f.write("State Dict:\n")
+        for key, value in state_dict.items():
+            f.write(f"{key}:\n")
+            f.write(f"  Shape: {value.shape}\n")
+            f.write(f"  Values: {value}\n")
+
+
+def format_size(size_bytes: int) -> str:
+    """Formats bytes into a human-readable size string."""
+    if size_bytes < 1024:
```

Comment:
I would recommend setting 1024 to a constant called MB (or something similar) and reuse that. Reduces chance of accidentally altering one of the magic numbers.

Reply:
Done